README.md
# Elastalert Hive Alerter

- [About](#About)
- [Requirements](#Requirements)
- [Installation](#Installation)
    * [For Elastalert 2](#For-Elastalert-2)
    * [As Responder for theHive](#As-Responder-for-theHive)
- [Configuration](#Configuration)
- [Additional Documentation](#Additional-Documentation)

## About
- This package is an enhancement addon to Elastalert 2 <https://github.com/jertel/elastalert2>
- It provides a custom `Elastalert Alerter`
<https://elastalert2.readthedocs.io/en/latest/recipes/adding_alerts.html#adding-a-new-alerter> which creates alerts with observables in TheHive <https://thehive-project.org/> using `TheHive4Py` <https://github.com/TheHive-Project/TheHive4py>.

- It provides two data contexts. The "rule" context provides information about the Elastalert rule,
eg. the rule name. The "match" context provides the data that the rule has matched.

- Data from either context can be used to configure the alert and / or to create data for an observable.

- The context data is specified via normal python string formatting (see examples below).

----

This package also provides a `custom Elastalert Enhancement` <https://elastalert2.readthedocs.io/en/latest/recipes/adding_enhancements.html> which will suppress alerts raised by the Alerter if a hash of the observables in the raised alert is found in the specified Elasticsearch database.

The hashes should be inserted into the database from another source, most likely the ObservableHashCreator `responder` <https://github.com/TheHive-Project/CortexDocs/blob/master/api/how-to-create-a-responder.md> in `CortexAnalyzers` <https://github.com/TheHive-Project/Cortex-Analyzers>

----

Note: It is possible to place static configuration such as *hive_connection* or *es_alert_hashes_connection* in the Elastalert config file instead of the rule file.

----


## Requirements

- Python 3.6 (Python 3.8 recommended)
- Python Pip
- Python Pipenv or your flavour of Python virtual environment.

## Installation
### For Elastalert 2

#### Python Virtual Environment

1. Create and change into directory for Elastalert.
2. Create Python virtual environment.
    ```bash
    pipenv --python3.8
    # activate environment
    pipenv shell
    ```
3. Clone Elastalert Hive Alerter
    ```bash
    git clone git@github.com:Nclose-ZA/elastalert_hive_alerter.git
    ```
4. Install. Please note this will install Elastalert2 as well as ElastAlert Hive Alerter.
    ```bash
    pip install elastalert_hive_alerter
    ```

If you already have Elastalert2 running you can just follow steps 3 and 4.

 
----

## As Responder for theHive
### Install TheHive responder (To suppress events from theHive web instance)

 1. Copy the `thehive_suppressor`to the approriate directory on your theHive instance (usually located in `/opt/Cortex-Analyzers/responders/ObservableHashCreator` on Docker deployments).
 2. Log into Cortex. Navigate to `Organization`, `Responders Config` and you should see the `ObservableHashCreator` if the above was followed correctly. Configure the necessary requirements.

----
## Configuration

1. Set the alerter in the rule file:
    ```yaml
    alert: "elastalert_hive_alerter.hive_alerter.HiveAlerter"
    ```

2. Configure connection details for TheHive (required fields shown first) in either the config file or the rule file:
    ```yaml
    hive_connection:
      hive_host: http(s)://sample_host
      hive_port: <hive_port>
      hive_apikey: <hive_apikey>

    hive_proxies:
      http: ''
      https: ''
    ```

3. Configure the alert by providing parameters consumed by TheHive4Py (required fields shown first) in the rule file:
    ```yaml
    hive_alert_config:
      title: 'Sample Title'  ## This will default to {rule[index]_rule[name]} if not provided
      type: 'external'
      source: 'instance1'
      description: '{match[field1]} {rule[name]} Sample description'

      severity: 2
      tags: ['sample_tag_1', 'sample_tag_2 {rule[name]}']
      tlp: 3
      status: 'New'
      follow: True
    ```

4. If desired, matched data fields can be mapped to TheHive observable types using python string formatting in the rule file:
    ```yaml
    hive_observable_data_mapping:
      - domain: "{match[field1]}_{rule[name]}"
      - domain: "{match[field]}"
      - ip: "{match[ip_field]}"
    ```

5. Set the enhancement in the rule file:
    ```yaml
    match_enhancements:
      - elastalert_hive_alerter.hive_alerter.HashSuppressorEnhancement
    ```

6. Configure connection details for Elasticsearch in either the config file or the rule file:
    ```yaml
    es_alert_hashes_connection:
      es_host: 'localhost'
      es_port: 9200
      es_username:
      es_password:
      index: 'alert_hashes'
      use_ssl:
      verify_certs:
      ca_certs:
      client_cert:
      client_key:
    ```

## Additional Documentation

https://elastalert2.readthedocs.io/en/latest/


CHANGELOG.md
# Elastalert Hive Alerter Changelog

## [Unreleased]

### Added
* none

### Changed
* none

### Removed
* none

----

## Elastalert Hive Alerter [1.1.1] - Date 5 May 2021

Updated and test to work with latest version of theHive4py version 1.8.1

### Added
* CHANGELOG.md
* README.md
* Removed Python 2.* as classifiers from setup.py as Python 2.* not supported.


### Removed
* Remove override JSONSerializable.jsonify class as it is no longer needed.
* README.rst

elastalert_hive_alerter/tests/test_alerter.py
# -*- coding: utf-8 -*-
from mock import patch
from elastalert.ruletypes import AnyRule
import time
from unittest import TestCase

a_date = int(time.time())*1000

rule = {
    'index': 'test_index',
    'name': 'test_rule_name',
    'hive_observable_data_mapping': [
        {'filename': '{match[domain]}_{match[ip_address]}.txt'},
        {'domain': '{match[domain]}'},
        {'domain': '{match[some_other_domain]}'},
        {'ip': '{match[ip_address]}'},
        {'domain': '{match[nested1][nested2]}'},
        # This should not result in an artifact as it is not present in the match
        {'other': '{match[SubjectUserName]}'}
    ],
    'hive_connection': {
        'hive_password': 'password',
        'hive_username': 'username',
        'hive_port': 9000,
        'hive_host': 'http://test_host',
        'hive_proxies': {'http': '', 'https': ''}
    },
    'hive_alert_config': {
        'status': 'New',
        'tlp': 3,
        'severity': 2,
        'tags': ['TheHive4Py', 'sample {rule[name]}'],
        'source': 'instance1',
        'follow': True,
        'type': 'external',
        'description': '{rule[name]}\n{match[domain]} Test desc',
        'customFields': {
            'test_date': {'type': 'date', 'value': '{match[a_date]}'},
            'test_string': {'type': 'string', 'value': '{match[some_other_domain]}'},
            'test_boolean': {'type': 'boolean', 'value': '{match[a_boolean]}'},
            'test_number': {'type': 'number', 'value': '{match[a_number]}'}
        }
    }
}

rule_classic = {
    'index': 'test_index',
    'name': 'test_rule_name',
    'type': AnyRule({}),
    'hive_observable_data_mapping': [
        {'filename': '{match[domain]}_{match[ip_address]}.txt'},
        {'domain': '{match[domain]}'},
        {'domain': '{match[some_other_domain]}'},
        {'ip': '{match[ip_address]}'},
        {'domain': '{match[nested1][nested2]}'}
    ],
    'hive_connection': {
        'hive_password': 'password',
        'hive_username': 'username',
        'hive_port': 9000,
        'hive_host': 'http://test_host',
        'hive_proxies': {'http': '', 'https': ''}
    },
    'hive_alert_config_type': 'classic',
    'hive_alert_config': {
        'status': 'New',
        'tlp': 3,
        'severity': 2,
        'tags': ['TheHive4Py', 'sample {rule[name]}'],
        'source': 'instance1',
        'follow': True,
        'type': 'external',
        'description': '{rule[name]}\n{match[domain]} Test desc',
        'customFields': {
            'test_date': {'type': 'date', 'value': '{match[a_date]}'},
            'test_string': {'type': 'string', 'value': '{match[some_other_domain]}'},
            'test_boolean': {'type': 'boolean', 'value': '{match[a_boolean]}'},
            'test_number': {'type': 'number', 'value': '{match[a_number]}'}
        }
    }
}

matches = [
    {
        'ip_address': '1.1.1.1',
        'domain': 'test.com',
        'some_other_domain': 'test2.com',
        'nested1': {'nested2': 'nested_value'},
        'a_boolean': True,
        'a_date': a_date,
        'a_number': 123
    }
]

matches_classic = [
    {
        'ip_address': '1.1.1.1',
        'domain': 'test.com',
        'some_other_domain': 'test2.com',
        'nested1': {'nested2': 'nested_value'},
        'a_boolean': True,
        'a_date': a_date,
        'a_number': 123
    },
    {
        'ip_address': '1.2.3.4',
        'domain': 'test.com',
        'some_other_domain': 'test2.com',
        'nested1': {'nested2': 'nested_value'},
        'a_boolean': True,
        'a_date': a_date,
        'a_number': 123
    }
]

class TestAlerter(TestCase):
    def test_alerter(self):
        class FakeResponse(object):
            status_code = 201
            text = ''

        class FakeApi(object):
            def create_alert(self, *args, **kwargs):
                return FakeResponse()

        with patch('thehive4py.api.TheHiveApi', return_value=FakeApi()) as mock_api, \
                patch.object(FakeApi, 'create_alert', return_value=FakeResponse()) as mock_create, \
                patch('thehive4py.models.Alert') as mock_alert, \
                patch('thehive4py.models.AlertArtifact', return_value='fake_artifact_return') as mock_artifact, \
                patch('uuid.uuid4', return_value='123456') as mock_uuid4:
            from elastalert_hive_alerter.hive_alerter import HiveAlerter
            hive_alerter = HiveAlerter(rule)
            hive_alerter.alert(matches)

            mock_api.assert_called_with(
                u'http://test_host:9000',
                u'',
                cert=False,
                proxies={u'http': u'', u'https': u''}
            )

            mock_artifact.assert_any_call(data=u'test.com_1.1.1.1.txt', dataType=u'filename')
            mock_artifact.assert_any_call(data=u'test.com', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'test2.com', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'nested_value', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'1.1.1.1', dataType=u'ip')
            mock_alert.assert_called_with(
                artifacts=[u'fake_artifact_return' for i in xrange(len(rule['hive_observable_data_mapping'])-1)],
                description=u'test_rule_name\ntest.com Test desc',
                follow=True,
                severity=2,
                source=u'instance1',
                sourceRef='123456',
                status=u'New',
                tags=[u'TheHive4Py', u'sample test_rule_name'],
                title=u'test_rule_name',
                tlp=3,
                type=u'external',
                customFields={
                    u'test_number': {'order': 0, 'number': '123'},
                    u'test_boolean': {'order': 1, 'boolean': u'True'},
                    u'test_date': {'order': 2, 'date': unicode(a_date)},
                    u'test_string': {'order': 3, 'string': u'test2.com'}
                }
            )

            mock_api.reset_mock()
            mock_artifact.reset_mock()
            mock_alert.reset_mock()

            hive_alerter = HiveAlerter(rule_classic)
            hive_alerter.alert(matches_classic)

            mock_api.assert_called_with(
                u'http://test_host:9000',
                u'',
                cert=False,
                proxies={u'http': u'', u'https': u''}
            )

            mock_artifact.assert_any_call(data=u'test.com_1.1.1.1.txt', dataType=u'filename')
            mock_artifact.assert_any_call(data=u'test.com', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'test2.com', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'nested_value', dataType=u'domain')
            mock_artifact.assert_any_call(data=u'1.1.1.1', dataType=u'ip')
            mock_alert.assert_called_with(
                artifacts=[u'fake_artifact_return' for i in xrange(len(rule_classic['hive_observable_data_mapping'])*2)],
                description=u'test_rule_name\n\na_boolean: True\na_date: '+unicode(a_date)+'\na_number: 123\ndomain: test.com\nip_address: 1.1.1.1\nnested1: {\n    "nested2": "nested_value"\n}\nsome_other_domain: test2.com\n\n----------------------------------------\ntest_rule_name\n\na_boolean: True\na_date: '+unicode(a_date)+'\na_number: 123\ndomain: test.com\nip_address: 1.2.3.4\nnested1: {\n    "nested2": "nested_value"\n}\nsome_other_domain: test2.com\n\n----------------------------------------\n',
                follow=True,
                severity=2,
                source=u'instance1',
                sourceRef='123456',
                status=u'New',
                tags=[u'TheHive4Py', u'sample test_rule_name'],
                title=u'test_rule_name',
                tlp=3,
                type=u'external',
                customFields={
                    u'test_number': {'order': 0, 'number': '123'},
                    u'test_boolean': {'order': 1, 'boolean': u'True'},
                    u'test_date': {'order': 2, 'date': unicode(a_date)},
                    u'test_string': {'order': 3, 'string': u'test2.com'}
                }
            )


thehive_suppressor/ObservableHashCreator.py
#!/usr/bin/env python
# encoding: utf-8

import datetime
import hashlib
import json
import sys

from cortexutils.responder import Responder
from elasticsearch_dsl import connections, Document, Keyword, Index, Date
from thehive4py.models import AlertArtifact, JSONSerializable, CustomJsonEncoder


# Monkey patch jsonify as the indent keyword results in different hashes between python 2 & 3
def jsonify(self):
    return json.dumps(self, sort_keys=True, cls=CustomJsonEncoder)
JSONSerializable.jsonify = jsonify


class ObservableHashCreator(Responder):
    """
    Creates an MD5 hash of the observables in the specified Elasticsearch database
    """

    def __init__(self):

        Responder.__init__(self)

        self.es_host = self.get_param('config.es_host', message='You have to configure an Elasticsearch host')
        self.es_port = self.get_param('config.es_port', default="9200")
        self.es_username = self.get_param('config.es_username')
        self.es_password = self.get_param('config.es_password')
        self.es_index = self.get_param('config.es_index', default='alert_hashes')

        self.es_kwargs = {
            'hosts': [':'.join([self.es_host, self.es_port])]
        }

        if self.es_username:
            self.es_kwargs['http_auth'] = (self.es_username, self.es_password)

        self.es_kwargs.update({
            'use_ssl': self.get_param('es_use_ssl', default=False),
            'verify_certs': self.get_param('es_verify_certs', default=False),
            'client_cert': self.get_param('es_client_cert'),
            'client_key': self.get_param('es_client_key'),
            'ca_certs': self.get_param('es_ca_certs'),
            'timeout': self.get_param('es_timeout', default=20)
        })

    def run(self):
        """
        Add a hash to the specified Elasticsearch index if some artifacts have been submitted and the hash does not
            already exist
        """

        Responder.run(self)

        data = self.get_data()
        artifacts = self.get_param('data.artifacts')
        rule_title = self.get_param('data.title', default='no data')
        suppressing_username = self.get_param('data.createdBy', default='no data')
        if not artifacts:  # If there are no artifacts we have no data to generate a hash
            self.error('No artifacts submitted')
        observables = []
        # The jsonify method provides a list of predictably sorted JSON strings which we then sort in order to make
        #   sure that the we generate the same hash that will be read from the database by HashSuppressorEnhancement
        for artifact in artifacts:
            observables.append(AlertArtifact(dataType=artifact.get('dataType'), data=artifact.get('data')))
        observable_hash_string = u'|'.join(sorted([observable.jsonify() for observable in observables]))
        observable_hash = hashlib.md5(observable_hash_string.encode('utf-8')).hexdigest()

        class AlertHash(Document):
            alert_hash = Keyword()
            date_suppressed = Date()
            rule_title = Keyword()
            suppressing_username = Keyword()
            class Index:
                name = self.es_index  # This is required because the commented registration below does not work

        connections.create_connection(**self.es_kwargs)
        alert_hashes = Index(self.es_index)
        # alert_hashes.document(AlertHash)  # Registering a Document to the Index does not work for some reason
        if not alert_hashes.exists():
            alert_hashes.create()
        results = AlertHash.search().filter('term', alert_hash=observable_hash).execute(ignore_cache=True)
        if not results:  # Don't save the same hash multiple times
            alert_hash = AlertHash(
                alert_hash=observable_hash,
                date_suppressed=datetime.datetime.now(),
                rule_title=rule_title,
                suppressing_username=suppressing_username
            )
            alert_hash.save()
            message = 'MD5 hash [{}] added to Elasticsearch database [{}:{}] index [{}]'.format(
                observable_hash, self.es_host, self.es_port, self.es_index)
        else:
            message = 'MD5 hash [{}] already existed in the Elasticsearch database [{}:{}] index [{}]'.format(
                observable_hash, self.es_host, self.es_port, self.es_index)

        self.report({'message': message})


if __name__ == '__main__':
    ObservableHashCreator().run()



elastalert_hive_alerter/hive_alerter.py
# -*- coding: utf-8 -*-
import hashlib
import logging
import re
import uuid
import json

from elastalert.alerts import Alerter
from elastalert.enhancements import BaseEnhancement, DropMatchException
from elasticsearch_dsl import connections, Document, Keyword, Index
from thehive4py.api import TheHiveApi
from thehive4py.models import Alert, AlertArtifact, CustomFieldHelper, JSONSerializable, CustomJsonEncoder


elastalert_logger = logging.getLogger('elastalert')


def _create_artifacts(rule, match):

    context = {'rule': rule, 'match': match}

    artifacts = []
    for mapping in rule.get('hive_observable_data_mapping', []):
        for observable_type, match_data_key in mapping.items():
            try:
                match_data_keys = re.findall(r'\{match\[([^\]]*)\]', match_data_key)
                rule_data_keys = re.findall(r'\{rule\[([^\]]*)\]', match_data_key)
                data_keys = match_data_keys + rule_data_keys
                context_keys = list(context['match'].keys()) + list(context['rule'].keys())
                if all([True if k in context_keys else False for k in data_keys]):
                    artifacts.append(AlertArtifact(dataType=observable_type, data=match_data_key.format(**context)))
            except KeyError as e:
                print('format string failed for key {}\nformat string\n{}\ncontext\n{}'.format(
                    e, match_data_key, context))
    return artifacts


def _create_alert_config(rule, match):

    context = {'rule': rule, 'match': match}

    alert_config = {
        'artifacts': _create_artifacts(rule, match),
        'sourceRef': str(uuid.uuid4())[0:6],
        'title': '{rule[name]}'.format(**context)
    }

    alert_config.update(rule.get('hive_alert_config', {}))
    
    for alert_config_field, alert_config_value in alert_config.items():
        if alert_config_field == 'customFields':
            custom_fields = CustomFieldHelper()
            for cf_key, cf_value in alert_config_value.items():
                try:
                    func = getattr(custom_fields, 'add_{}'.format(cf_value['type']))
                except AttributeError:
                    raise Exception('unsupported custom field type {}'.format(cf_value['type']))
                value = cf_value['value'].format(**context)
                func(cf_key, value)
            alert_config[alert_config_field] = custom_fields.build()
        elif isinstance(alert_config_value, str):
            alert_config[alert_config_field] = alert_config_value.format(**context)
        elif isinstance(alert_config_value, (list, tuple)):
            formatted_list = []
            for element in alert_config_value:
                try:
                    formatted_list.append(element.format(**context))
                except (AttributeError, KeyError, IndexError):
                    formatted_list.append(element)
            alert_config[alert_config_field] = formatted_list

    return alert_config


class HashSuppressorEnhancement(BaseEnhancement):

    required_options = set(['es_suppression_hashes_connection'])  # This is not currently used, just informational

    def process(self, match):
        """
        This Elastalert enhancement will suppress an alert raise by HiveAlerter if a matching hash is found in the
            specified database
        It was written to be used in conjunction with the ObservableHashCreator responder for TheHive
        """
        
        connection_details = self.rule['es_alert_hashes_connection']

        kwargs = {}

        if connection_details.get('es_host'):
            kwargs['hosts'] = ['{}:{}'.format(connection_details['es_host'], connection_details.get('es_port', 9200))]

        if connection_details.get('es_username'):
            kwargs['http_auth'] = (connection_details['es_username'], connection_details['es_password'])

        kwargs.update({
            'use_ssl': connection_details.get('use_ssl', False),
            'verify_certs': connection_details.get('verify_certs', False),
            'client_cert': connection_details.get('client_cert', None),
            'client_key': connection_details.get('client_key', None),
            'ca_certs': connection_details.get('ca_certs', None),
            'timeout': connection_details.get('es_conn_timeout', 20)
        })

        class AlertHash(Document):
            alert_hash = Keyword()

        connections.create_connection(**kwargs)
        alert_hashes = Index(connection_details.get('index', 'alert_hashes'))
        alert_hashes.document(AlertHash)
        if not alert_hashes.exists():
            alert_hashes.create()

        alert_config = _create_alert_config(self.rule, match)
        # The jsonify method provides a list of predictably sorted JSON strings which we then sort in order to make
        #   sure that the we generate the same hash that was written to the database by ObservableHashCreator
        observables = sorted([observable.jsonify() for observable in alert_config['artifacts']])
        observable_hash_string = u'|'.join(observables)
        observable_hash = hashlib.md5(observable_hash_string.encode('utf-8')).hexdigest()
        results = AlertHash.search().filter('term', alert_hash=observable_hash).execute(ignore_cache=True)

        if results:
            elastalert_logger.info('Alert was not sent because hash [{}] was found'.format(observable_hash))
            raise DropMatchException()
        else:
            elastalert_logger.info('Alert was sent because hash [{}] was not found'.format(observable_hash))


class HiveAlerter(Alerter):
    """
    Use matched data to create alerts containing observables in an instance of TheHive
    """

    required_options = set(['hive_connection', 'hive_alert_config'])

    def get_aggregation_summary_text(self, matches):
        text = super(HiveAlerter, self).get_aggregation_summary_text(matches)
        if text:
            text = '```\n{0}```\n'.format(text)
        return text

    def send_to_thehive(self, alert_config):
        connection_details = self.rule['hive_connection']
        api = TheHiveApi(
            '{hive_host}:{hive_port}'.format(**connection_details),
            connection_details.get('hive_apikey', ''),
            proxies=connection_details.get('hive_proxies', {'http': '', 'https': ''}),
            cert=connection_details.get('hive_verify', False))

        alert = Alert(**alert_config)
        try:
            response = api.create_alert(alert)
        except Exception as e:
            elastalert_logger.warning("Error connecting to theHive. ERROR: ".format(e))

        if response.status_code != 201:
            raise Exception('alert not successfully created in TheHive\n{}'.format(response.text))

    def alert(self, matches):
        if self.rule.get('hive_alert_config_type', 'custom') != 'classic':
            for match in matches:
                alert_config = _create_alert_config(self.rule, match)
                self.send_to_thehive(alert_config)
        else:
            alert_config = _create_alert_config(self.rule, matches[0])
            artifacts = []
            for match in matches:
                artifacts += _create_artifacts(self.rule, match)
                if 'related_events' in match:
                    for related_event in match['related_events']:
                        artifacts += _create_artifacts(self.rule, related_event)

            alert_config['artifacts'] = artifacts
            alert_config['title'] = self.create_title(matches)
            alert_config['description'] = self.create_alert_body(matches)
            self.send_to_thehive(alert_config)

    def get_info(self):

        return {
            'type': 'hivealerter',
            'hive_host': self.rule.get('hive_connection', {}).get('hive_host', '')
        }
